extensions/multiply-divide/README.md

```diff
@@ -0,0 +1,50 @@
+# General design
+
+**Extension State: Under Development**  
+**Requires: Base, VWI, CP2.0**  
+**CPUID Bit: CP2.3**
+
+# Overview
+
+This extension adds multiplication, division, and remainder instructions. 
+
+# Added Instructions
+
+These instructions are in the expanded calculation opcode section of instructions.
+
+A `U` at the front of an instruction indicates that it treats its operands as unsigned.
+An `S` indicates that the operands are treated as signed.
+
+### Opcode table
+
+| `C CCCC CCCC` | NAME     | Operation                                  | Flags  | Comment |
+|---------------|----------|--------------------------------------------|--------|---------|
+| `0 0001 0000` | `UDIV`   | `A ← A / B`                                | None   | (1)     |
+| `0 0001 0001` | `SDIV`   | `A ← A / B`                                | None   | (1)     |
+| `0 0001 0010` | `UREM`   | `A ← A % B`                                | None   | (1) (2) |
+| `0 0001 0011` | `SREM`   | `A ← A % B`                                | None   | (1) (2) |
+| `0 0001 0100` | `UMUL`   | `A ← A × B`                                | `CV`   | (3) (5) |
+| `0 0001 0101` | `SMUL`   | `A ← A × B`                                | `CV`   | (3) (6) |                                          
+| `0 0001 0110` | `UMULH`  | `A ← high(A × B)`                          | None   | (4)     |
+| `0 0001 0111` | `SMULH`  | `A ← high(A × B)`                          | None   | (4)     |                                          
```

Comment:
Is there a reason for not defining the Z and N bits for these operations? Also, why define BOTH C and V flags if they're going to be the same?

Reply:
Mostly matching x86. Also see above about flags.

For ARM, depending on the flavor, either no flags or set, or the Z and N flags are set; C is untouched (or corrupted on old versions) and V is untouched. We could set all of them, I'm not opposed to that. But leaving flags untouched is specifically against part of our rationale.

Reply:
I 100% agree that mandating unmentioned flags be untouched isn't something that we should do. I think it makes sense for the Z and N flags to be set with these instructions. I don't think we need to set BOTH C and V. Only setting one of them should be good enough.

Reply:
This also brings up the possibility of having ALU instructions which leave ALL the flags in undefined states rather than preserving all of them. We should probably have some sort of way to differentiate them if both of those types of instructions are possible.

Reply:
I'm not opposed to setting ZNC. I'd like to find out why ARM doesn't set C based on the upper half, but I wouldn't know who to ask. Maybe there's some former ARM developer on stackoverflow lol.

I'm not sure what the benefit of such an ALU instruction would be.

Reply:
Thinking about common uses of division and remainder, aren't comparing both against 0 pretty common as well as checking if the remainder is negative (for implementing modulus)? Maybe we want to set the negative flag for signed division and remainder and the zero flag for both divisions and remainders?

Reply:
I agree that that could be useful. Good litmus test for flags: what does ARM do? Unfortunately, it's very hard to find _justifications_ for anything ARM does... I swear it feels like I need a manual for how to read their manuals.

Reply:
I just looked at the ARM manual and it says that `MUL` updates the `N` and `Z` flags. It also will corrupt or preserve the `C` and `V` flags depending on the version of ARM. Interestingly enough, the HMUL instructions don't affect the flags at all on ARM.

For the division instructions, ARM doesn't touch the conditions.

Reply:
Suggested changes are below as this is considered out-of-date by github